docs/types/reference-types.rst

```diff
@@ -468,6 +468,117 @@ Array Members
         }
     }
 
+.. index:: ! array;dangling storage references
+
+Dangling References to Storage Array Elements
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When working with storage arrays, you need to take care to avoid dangling references.
+A dangling reference is a reference that points to something that no longer exists or has been
+moved without updating the reference. A dangling reference can for example occur, if you store a
+reference to an array element in a local variable and then ``.pop()`` from the containing array:
+
+.. code-block:: solidity
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity >=0.8.0 <0.9.0;
+
+    contract C {
+        uint[][] s;
+
+        function f() public {
+            // Stores a pointer to the last array element of s.
+            uint[] storage ptr = s[s.length - 1];
+            // Removes the last array element of s.
+            s.pop();
+            // Writes to the array element that is no longer at a valid index in s.
+            ptr.push(0x42);
+            // Adding a new element to ``s`` now will not add an empty array, but
+            // will result in an array of length 1 with ``0x42`` as element.
+            s.push();
+            assert(s[s.length - 1][0] == 0x42);
```

Comment:
This part feels like it comes out of nowhere since the text before only mentions dangling references and this feels like a separate issue - even if you understand dangling references you might not be aware that `push()` does not zero memory.

I think that we should first give an example showing that a dangling reference can let you write outside of the array and then, separately, show another one with `push()` to give it the proper emphasis.

But since we're pressed for time, I think this is something that could be changed after the release rather than in this PR.

Reply:
Is it? It's one of the more serious consequences but I think that dangling references would be problem even if `push()` was zeroing storage.